manubot/pubmed.py

```diff
@@ -0,0 +1,151 @@
+        'id': pmid,
+        'rettype': 'full',
+    }
+    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi'
+    response = requests.get(url, params)
+    try:
+        element_tree = xml.etree.ElementTree.fromstring(response.text)
+        element_tree, = element_tree.findall('PubmedArticle')
+    except Exception as error:
+        logging.error(f'Error fetching PubMed metadata for {pmid}.\n'
+                      f'Invalid response from {response.url}:\n{response.text}')
+        raise error
+    citeproc = citeproc_from_pubmed_article(element_tree)
+    return citeproc
+
+
+def citeproc_from_pubmed_article(article):
+    """
+    article is a PubmedArticle xml element tree
+    
+    https://github.com/citation-style-language/schema/blob/master/csl-data.json
+    """
+    citeproc = collections.OrderedDict()
+
+    title = article.findtext("MedlineCitation/Article/ArticleTitle")
+    if title:
+        citeproc['title'] = title
+
+    volume = article.findtext("MedlineCitation/Article/Journal/JournalIssue/Volume")
+    if volume:
+        citeproc['volume'] = volume
+
+    issue = article.findtext("MedlineCitation/Article/Journal/JournalIssue/Issue")
+    if issue:
+        citeproc['issue'] = issue
+
+    page = article.findtext("MedlineCitation/Article/Pagination/MedlinePgn")
+    if page:
+        citeproc['page'] = page
+
+    journal = article.findtext("MedlineCitation/Article/Journal/Title")
+    if journal:
+        citeproc['container-title'] = journal
+
+    journal_short = article.findtext("MedlineCitation/Article/Journal/ISOAbbreviation")
+    if journal_short:
+        citeproc['container-title-short'] = journal_short
+
+    issn = article.findtext("MedlineCitation/Article/Journal/ISSN")
+    if issn:
+        citeproc['ISSN'] = issn
+
+    date_parts = extract_publication_date_parts(article)
+    if date_parts:
+        citeproc['issued'] = {'date-parts': [date_parts]}
+
+    authors_csl = list()
+    authors = article.findall("MedlineCitation/Article/AuthorList/Author")
+    for author in authors:
+        author_csl = collections.OrderedDict()
+        given = author.findtext('ForeName')
+        if given:
+            author_csl['given'] = given
+        family = author.findtext('LastName')
+        if family:
+            author_csl['family'] = family
+        authors_csl.append(author_csl)
+    if authors_csl:
+        citeproc['author'] = authors_csl
+
+    for id_type, key in ('pubmed', 'PMID'), ('pmc', 'PMCID'), ('doi', 'DOI'):
+        xpath = f"PubmedData/ArticleIdList/ArticleId[@IdType='{id_type}']"
+        value = article.findtext(xpath)
+        if value:
+            citeproc[key] = value.lower() if key == 'DOI' else value
+
+    abstract = article.findtext("MedlineCitation/Article/Abstract/AbstractText")
+    if abstract:
+        citeproc['abstract'] = abstract
+
+    citeproc['URL'] = f"https://www.ncbi.nlm.nih.gov/pubmed/{citeproc['PMID']}"
+    citeproc['type'] = 'article-journal'
```

Comment:
Is `article-journal` always the article type?  PubMed apparently has [books](https://www.nlm.nih.gov/bsd/disted/pubmedtutorial/010_070.html) as well.  As an example, https://www.ncbi.nlm.nih.gov/pubmed/29227604.  It looks like a few other things like the `title` (`BookTitle` in the XML) will fail for books.

I'm okay leaving this as an uncovered corner case, but we may want to document that.

Reply:
Good catch. Previously this would fail. But now I made the failure better documented and have it throw a `NotImplementedError`. Added a test case